resolwe/flow/models/entity.py

```diff
@@ -58,27 +59,59 @@ class Meta(BaseCollection.Meta):
             ("owner_relation", "Is owner of the relation"),
         )
 
+    UNIT_HOUR = 'hr'
+    UNIT_CHOICES = (
+        (UNIT_HOUR, 'Hour'),
+    )
+
+    def validate_unique(self, exclude=None):
+        """Check unique constraints on the model.
+
+        Separate check for the combination of ``collection`` and ``category``
+        is required as ``category`` is optional and thus
+        ``Meta.unique_together`` cannot be used.
+        """
+        super().validate_unique(exclude)
+
+        duplicate_exists = Relation.objects.exclude(id=self.id).filter(
+            collection_c=self.collection, category=self.category, category__isnull=False
+        ).exists()
+
+        if duplicate_exists:
+            raise ValidationError("Combination of 'collection' and 'category' must be unique.")
+
     #: type of the relation
     type = models.ForeignKey('RelationType', on_delete=models.PROTECT)
 
     #: collection in which relation is
     collection = models.ForeignKey('Collection', null=True, on_delete=models.SET_NULL)
 
-    #: entities in the relation
-    entities = models.ManyToManyField(Entity, through='PositionInRelation')
+    #: enteties in the relation
+    entities = models.ManyToManyField(Entity, through='RelationPartition')
 
     #: optional label of the relation (i.e. to further specify relation type)
-    label = models.CharField(max_length=100, null=True, blank=True)
+    category = models.CharField(max_length=100, null=True, blank=True)
+
+    #: unit used in the partitions
+    partition_unit = models.CharField(max_length=2, choices=UNIT_CHOICES, null=True, blank=True)
 
 
-class PositionInRelation(models.Model):
+class RelationPartition(models.Model):
     """Intermediate model to connect Entities with Relations."""
 
+    class Meta:
+        """RelationPartition Meta options."""
+
+        unique_together = ("entity", "relation")
+
     #: entity in the relation
     entity = models.ForeignKey(Entity, on_delete=models.CASCADE)
 
     #: relation to which entity belongs
     relation = models.ForeignKey(Relation, on_delete=models.CASCADE)
 
-    #: position of the entity in the relation (relevant if `relation.type.ordered` is True)
-    position = JSONField(null=True, blank=True)
+    #: position of the entity in the relation (relevant if `relation.type.odrered` is True)
+    position = models.PositiveSmallIntegerField(null=True, blank=True, db_index=True)
```

Comment:
I hope `PositiveSmallIntegerField` will cover all the use cases. I agree we start with that and change it only if there is a viable use case.